packages/react-router/lib/server-runtime/data.ts

```diff
@@ -20,60 +22,37 @@ export interface AppLoadContext {
  */
 export type AppData = unknown;
 
-export async function callRouteAction({
-  loadContext,
-  action,
-  params,
-  request,
-  routeId,
-}: {
-  request: Request;
-  action: ActionFunction;
-  params: ActionFunctionArgs["params"];
-  loadContext: AppLoadContext;
-  routeId: string;
-}) {
-  let result = await action({
-    request: stripRoutesParam(stripIndexParam(request)),
-    context: loadContext,
-    params,
-  });
-
-  if (result === undefined) {
-    throw new Error(
-      `You defined an action for route "${routeId}" but didn't return ` +
-        `anything from your \`action\` function. Please return a value or \`null\`.`
+function checkRedirect(result: ReturnType<LoaderFunction | ActionFunction>) {
+  if (
+    isDataWithResponseInit(result) &&
+    result.init &&
+    isRedirectStatusCode(result.init.status || 200)
+  ) {
+    throw new Response(
+      new Headers(result.init.headers).get("Location")!,
+      result.init
     );
   }
-
-  return result;
 }
 
-export async function callRouteLoader({
+export async function callRouteHandler({
   loadContext,
-  loader,
+  handler,
   params,
   request,
-  routeId,
 }: {
   request: Request;
-  loader: LoaderFunction;
-  params: LoaderFunctionArgs["params"];
+  handler: LoaderFunction | ActionFunction;
+  params: LoaderFunctionArgs["params"] | ActionFunctionArgs["params"];
   loadContext: AppLoadContext;
-  routeId: string;
 }) {
-  let result = await loader({
+  let result = await handler({
     request: stripRoutesParam(stripIndexParam(request)),
     context: loadContext,
     params,
   });
 
-  if (result === undefined) {
-    throw new Error(
-      `You defined a loader for route "${routeId}" but didn't return ` +
-        `anything from your \`loader\` function. Please return a value or \`null\`.`
-    );
-  }
```

Comment:
I thought we had already removed this restriction for v7 - but that may have only happened in the router and not in this Remix wrapper. I think there's actually a decent bit of code consolidation we can do between `routes.ts` and `data.ts` that I may do in this PR when I get back from vaca.  

I don't see this fix as super critical since folks who want to redirect should be using `return/throw redirect()` - this just handles the odd case that you use `unstable_data` to give it a 3xx status code and a `Location` header